pages/review/[lesson].tsx

```diff
@@ -60,9 +60,9 @@ const Review: React.FC<QueryDataProps<AppData>> = ({ queryData }) => {
     router.push('/curriculum')
     return <LoadingSpinner />
   }
-  const lessonSubmissions: SubmissionData[] = data
+  const lessonSubmissions: Submission[] = data
```

Comment:
This type should also get inferred from the query type.

Also the `Submission` type might not always be the same as the one from the query, as the query may not include all the fields in the type. For this there's the autogenerated query result type: `SubmissionsQuery['submissions']`

Reply:
I tried to type incoming data but it became much less readable because you need either check for every optional field or assert non-nulls. We should fix our type definitions first. 

And I'm not quite sure if you need to type every query result like this `SubmissionsQuery['submissions']`. Codegen generates nullable fields like this `<Maybe T> | undefined` so you can safely use general `Submission` type for everything.

Reply:
That's not the issue that may arise. The generated types for the query still comes from the ones on the typedef, so the optional fields still gonna be there. That's only gonna get fixed by correcting the typedef.

Let me give you an example of an issue that this may cause. Imagine if now you needed the reviewer email from a submission for some reason, to do something in one of the components for the page, like to put a button to email the reviewer. Since the component uses the [`Submission`](https://github.com/garageScript/c0d3-app/blob/1dab2746ba0409f3f87b21930babec3894547398/graphql/typeDefs.ts#L94) type, which has a `reviewer` field of type [`User`](https://github.com/garageScript/c0d3-app/blob/1dab2746ba0409f3f87b21930babec3894547398/graphql/typeDefs.ts#L119), all the fields from `User` are available for you from Typescript. So you just get the email from there and that's it, it should work as expected. So you commit, open the PR and then it get's merged.

Then you push it to production and it doesn't work. You check the logs and there's a bunch of undefined value errors. But why though? You coded the new feature and all the tests and type checks were ok. What happened is that you didn't check the GraphQL query, and didn't see that the reviewer field only includes the `id` and `username`. Even though Typescript was telling you that everything was ok, in reality that wasn't true. Unfortunately this is a weakness of duck-typed languages like TS.

Of course this doesn't cause an immediate problem for this PR. What I'm trying to emphasize is that it's a code smell, that may lead someone not familiar with the code in the future to create a new bug.

Now, since the `submissions` query return type is still optional, until that's fixed you need to do a workaround to use the generated query type. Something like this:

```ts
type submissionData = NonNullable<SubmissionsQuery['submissions']>[number]
```

In the case of the code that this comment is referencing, the type of `data` is `any` because the component is not using the generated `useSubmissionsQuery` hook. If you use the correct typed hook, the types can be inferred very easily and typing will be much easier.

Reply:
I agree with the importance of proper typing but I still can't see how using general codegen types could cause any issues. Codegen is smart it doesn't use general types, it picks only fields which are defined in query. So with your reviewer email example, this is SubmissionQuery type:
```js
export type SubmissionsQuery = { __typename?: 'Query' } & {
  submissions?: Maybe<
    Array<
      Maybe<
        { __typename?: 'Submission' } & Pick<
          Submission,
          | 'id'
          | 'status'
          | 'diff'
          | 'comment'
          | 'challengeId'
          | 'lessonId'
          | 'createdAt'
          | 'updatedAt'
        > & {
            challenge?: Maybe<
              { __typename?: 'Challenge' } & Pick<Challenge, 'title'>
            >
            user?: Maybe<
              { __typename?: 'User' } & Pick<User, 'id' | 'username'>
            >
            reviewer?: Maybe<
              { __typename?: 'User' } & Pick<User, 'id' | 'username' | 'name'>
            >
          }
      >
    >
  >
}
```
You can't just type `submission.reviewer.email`, typescript will complain (a lot). First `submission` itself could be undefined or null, second `reviewer` can be undefined or null and according to this type `reviewer` doesn't even has `email` field. While general `User` type has it SubmissionQuery type is generated according to graphql query so it's `Pick<GeneralType, 'certain fields'>`.

Reply:
> SubmissionQuery type is more specific

Exactly, it's only a subset of the fields from `Submission` that are included on the query. The example I gave you was assuming you were using the more general `Submission` type.

As I said before, the optional fields are still gonna be the same, as they are inherited from the typedef. So null checks won't be affected by this.